src/components/Filter/useFiltersFromUrl.ts

```diff
@@ -0,0 +1,54 @@
+import { useMemo, useState } from 'react';
+import { useSearchParams } from '_/hooks/useSearchParams';
+
+import { Field } from '../types';
+
+import { GlobalFilters } from './types';
+
+const safeParse = (str: string) => {
+  try {
+    return JSON.parse(str);
+  } catch (e) {
+    return undefined;
+  }
+};
+
+export const useFiltersFromUrl = ({
+  fields,
+  filterPrefix = '',
+}: {
+  fields: Field[];
+  filterPrefix?: string;
+}): [GlobalFilters, (filters: GlobalFilters) => void] => {
+  const [searchParams, updateSearchParams] = useSearchParams();
+  const [selectedFilters, setSelectedFilters] = useState(() =>
+    Object.fromEntries(
+      fields
+        .map(({ id }) => ({
+          id,
+          params: safeParse(searchParams[`${filterPrefix}${id}`]),
+        }))
+        // deep check is not possible
+        // each filter should validate if values make sense (i.e. in range)
+        .filter(({ params }) => Array.isArray(params) && params.length)
+        .map(({ id, params }) => [id, params]),
+    ),
+  );
+  const setStateAndUrl = useMemo(
```

Comment:
Question - why do the `useMemo` to cache the set function?  I'm not sure why this is needed.

Reply:
I think that by convention, the objects returned from hooks should have stable references to prevent triggering re-render. In this case we would need to return an anonymous wrapper which would change with every invocation.

Reply:
That makes sense.